File: mathverse-miniapp/src/stores/learn.ts
import { create } from 'zustand';
import { request } from '../services/api';

interface KgNode {
  id: string;
  name: string;
  difficulty: number;
  mastery?: number;
}

interface KgSubject {
  id: string;
  name: string;
  chapters: {
    id: string;
    name: string;
    frequency: string;
    topics: KgNode[];
  }[];
}

interface Question {
  type: string;
  question: string;
  options?: string[];
  answer: string;
  analysis?: string;
}

interface LearnStore {
  stage: string;
  kgData: KgSubject[] | null;
  selectedKp: KgNode | null;
  lecture: string;
  exercises: Question[];
  grades: Record<number, { correct: boolean; feedback: string }>;
  addedMistakes: Record<number, boolean>;
  gradingIndex: number | null;
  activeTab: 'lecture' | 'exercise';
  isLoading: boolean;
  setStage: (stage: string) => void;
  fetchKg: () => Promise<void>;
  selectKp: (kp: KgNode) => void;
  fetchLecture: (kpName: string, kpId: string) => Promise<void>;
  fetchExercise: (kpName: string, kpId: string) => Promise<void>;
  gradeAnswer: (index: number, userAnswer: string) => Promise<void>;
  addExerciseMistake: (index: number, userAnswer: string) => Promise<void>;
}

const _subjectOf = (kpId: string): string => {
  if (kpId.startsWith('xd')) return '线性代数';
  if (kpId.startsWith('gl')) return '概率论与数理统计';
  if (kpId.startsWith('gs')) return '高等数学';
  return '';
};

export const useLearnStore = create<LearnStore>((set, get) => ({
  stage: 'college',
  kgData: null,
  selectedKp: null,
  lecture: '',
  exercises: [],
  grades: {},
  addedMistakes: {},
  gradingIndex: null,
  activeTab: 'lecture',
  isLoading: false,

  setStage: (stage) => set({ stage, selectedKp: null, lecture: '', exercises: [], grades: {}, addedMistakes: {} }),

  fetchKg: async () => {
    set({ isLoading: true });
    try {
      const data = await request<{ subjects: KgSubject[] }>(
        `/api/learn/kg/${get().stage}`
      );
      set({ kgData: data.subjects, isLoading: false });
    } catch {
      set({ isLoading: false });
    }
  },

  selectKp: (kp) => set({ selectedKp: kp, lecture: '', exercises: [], grades: {}, addedMistakes: {}, activeTab: 'lecture' }),

  fetchLecture: async (kpName, kpId) => {
    set({ isLoading: true });
    try {
      const data = await request<{ lecture: string }>('/api/learn/lecture', {
        method: 'POST',
        data: { kp_name: kpName, kp_id: kpId, stage: get().stage },
      });
      set({ lecture: data.lecture, isLoading: false, activeTab: 'lecture' });
    } catch {
      set({ isLoading: false });
    }
  },

  fetchExercise: async (kpName, kpId) => {
    set({ isLoading: true, activeTab: 'exercise', exercises: [], grades: {}, addedMistakes: {}, gradingIndex: null });
    try {
      const data = await request<{ questions: Question[] }>('/api/learn/exercise/generate', {
        method: 'POST',
        data: { kp_name: kpName, kp_id: kpId, stage: get().stage, count: 3 },
      });
      set({ exercises: data.questions, isLoading: false });
    } catch {
      set({ isLoading: false });
    }
  },

  gradeAnswer: async (index, userAnswer) => {
    const q = get().exercises[index];
    const kp = get().selectedKp;
    if (!q || !kp) return;
    set({ gradingIndex: index });
    try {
      const data = await request<{ correct: boolean; feedback: string }>('/api/learn/exercise/grade', {
        method: 'POST',
        data: {
          kp_id: kp.id,
          question: q.question,
          user_answer: userAnswer,
          reference_answer: q.answer,
        },
      });
      set((s) => ({
        grades: { ...s.grades, [index]: { correct: data.correct, feedback: data.feedback } },
        gradingIndex: null,
      }));
    } catch (e: any) {
      const feedback = e?.message === 'Authentication required' ? '请先登录后再判分' : '判分失败，请重试';
      set((s) => ({
        grades: { ...s.grades, [index]: { correct: false, feedback } },
        gradingIndex: null,
      }));
    }
  },

  addExerciseMistake: async (index, userAnswer) => {
    const q = get().exercises[index];
    const kp = get().selectedKp;
    if (!q || !kp || get().addedMistakes[index]) return;
    await request('/api/me/mistakes', {
      method: 'POST',
      data: {
        question_text: q.question,
        user_answer: userAnswer,
        correct_answer: q.answer,
        solution_steps: q.analysis ? [{ title: '解析', content: q.analysis }] : undefined,
        knowledge_point_id: kp.id,
        subject: _subjectOf(kp.id),
        difficulty: kp.difficulty,
      },
    });
    set((s) => ({ addedMistakes: { ...s.addedMistakes, [index]: true } }));
  },
}));

File: mathverse-api/app/routes/learn.py
"""学习 API routes — knowledge graph, lectures, exercises."""
import copy
import json
import re
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy.orm import Session
from app.middleware.auth_middleware import get_optional_user, get_current_user
from app.models.all import User, LearningProgress
from app.services.agent_client import agent_client, AgentUnavailableError
from app.services.deeptutor import tenancy
from app.services.kg import load_kg
from app.services.progress import record_attempt
from app.services.analytics import log_event
from app.database import get_db

router = APIRouter(prefix="/api/learn", tags=["learn"])


@router.get("/kg/{stage}")
async def get_knowledge_graph(
    stage: str,
    user: User | None = Depends(get_optional_user),
    db: Session = Depends(get_db),
):
    """Get knowledge graph for a stage, with user mastery data."""
    kg = load_kg(stage)

    # Inject user mastery data if available. Deep-copy first so the per-user
    # mastery never mutates the shared cache (else a logged-out request would
    # leak the previous user's mastery).
    if user:
        kg = copy.deepcopy(kg)
        records = db.query(LearningProgress).filter(
            LearningProgress.user_id == user.id
        ).all()
        progress_map = {r.knowledge_point_id: r.mastery_level for r in records}

        for subject in kg.get("subjects", []):
            for chapter in subject.get("chapters", []):
                for topic in chapter.get("topics", []):
                    topic["mastery"] = progress_map.get(topic["id"], 0.0)

    return kg


class LectureRequest(BaseModel):
    kp_name: str
    kp_id: str
    stage: str = "college"


@router.post("/lecture")
async def generate_lecture(req: LectureRequest, user: User | None = Depends(get_optional_user)):
    """Generate AI lecture for a knowledge point."""
    try:
        lecture = await agent_client.generate_lecture(req.kp_name, req.stage)
    except AgentUnavailableError as e:
        raise HTTPException(status_code=503, detail=f"AI讲课服务暂时不可用: {e}")
    return {"kp_id": req.kp_id, "lecture": lecture}


class ExerciseRequest(BaseModel):
    kp_id: str
    kp_name: str
    count: int = 5
    stage: str = "college"
    # RAG grounding (opt-in), same semantics as /api/solve: use_rag without
    # kb_name falls back to the shared curriculum library for the stage.
    use_rag: bool = False
    kb_name: str | None = None


@router.post("/exercise/generate")
async def generate_exercise(req: ExerciseRequest, user: User | None = Depends(get_optional_user)):
    """Generate practice exercises for a knowledge point."""
    if req.use_rag and req.kb_name and user:
        kb_name, enable_rag = tenancy.scope(user.id, req.kb_name), True
    elif req.use_rag:
        kb_name, enable_rag = tenancy.scope("curriculum", req.stage), True
    else:
        kb_name, enable_rag = None, False
    try:
        questions = await agent_client.generate_quiz(
            req.kp_name, req.count, req.stage, kb_name=kb_name, enable_rag=enable_rag,
        )
    except AgentUnavailableError as e:
        raise HTTPException(status_code=503, detail=f"出题服务暂时不可用: {e}")
    return {"kp_id": req.kp_id, "questions": questions}


class GradeRequest(BaseModel):
    kp_id: str
    question: str
    user_answer: str
    reference_answer: str | None = None


def _parse_grade(raw: str) -> dict:
    """Extract {correct, feedback} from the grader's reply, tolerating extra prose."""
    match = re.search(r"\{.*\}", raw, re.S)
    if match:
        try:
            data = json.loads(match.group(0))
            return {"correct": bool(data.get("correct")), "feedback": str(data.get("feedback", ""))}
        except json.JSONDecodeError:
            pass
    # Heuristic fallback when the model didn't return clean JSON.
    correct = "正确" in raw and "不正确" not in raw and "错误" not in raw
    return {"correct": correct, "feedback": raw.strip()[:200]}


@router.post("/exercise/grade")
async def grade_exercise(
    req: GradeRequest,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Grade a user's answer to a practice question and update mastery.

    Judging now goes through DeepTutor's question/judge WS (was a DeepSeek-direct
    bypass). The engine returns prose feedback; _parse_grade derives correctness.
    """
    try:
        raw = await agent_client.judge(
            req.question, req.user_answer, correct_answer=req.reference_answer
        )
    except AgentUnavailableError as e:
        raise HTTPException(status_code=503, detail=f"判题服务暂时不可用: {e}")
    verdict = _parse_grade(raw)
    record_attempt(db, user.id, req.kp_id, verdict["correct"])
    log_event(db, "exercise_graded", user.id, {"kp": req.kp_id, "correct": verdict["correct"]})
    return {"kp_id": req.kp_id, **verdict}
